metaverse_components/pet.js

```diff
@@ -90,45 +141,106 @@ export default (app, component) => {
+
+  useFrame(({ timestamp, timeDiff }) => {
+    const nowS = timestamp / 1000;
+    const timeDiffS = timeDiff / 1000;
+    const timeDiffCapped = Math.min(Math.max(timeDiff, 0), 100);
+    const timeDiffSCapped = timeDiffCapped / 1000;
+
+    if (fallLoopStartTimeS == 0) {
+      fallLoopStartTimeS = nowS;
+      startFallingPosition.copy(app.position);
+      localVector.set(0, 0, 0);
+    }
+    const t = nowS - fallLoopStartTimeS;
+    const h = 0.5 * physicsScene.getGravity().y * t * t;
+
+    const displacement = localVector.set(0, h - lastGravityH, 0);
+    smoothVelocity.add(displacement);
+    const flags = physicsScene.moveCharacterController(
+      characterController,
+      smoothVelocity,
+      0,
+      timeDiffSCapped,
+      app.position
+    );
+
+    lastGravityH = h;
+
+    let grounded = !!(flags & 0x1);
+    if (grounded) {
+      fallLoopStartTimeS = 0;
+      startFallingPosition.set(0, 0, 0);
+      lastGravityH = 0;
+    }
+
     // components
     const _updateAnimation = () => {
-      // const petComponent = app.getComponent('pet');
+      // making sure the nose of the pet hits the physics object in front of it
+      app.position.x -= (PET_WIDTH / 2) * lastDirection.x;
+      app.position.z -= (PET_WIDTH / 2) * lastDirection.z;
+
       if (petComponent) {
         if (rootBone) {
           rootBone.quaternion.copy(rootBone.originalQuaternion);
           rootBone.updateMatrixWorld();
         }
-        if (petMixer) { // animated pet
-          if (petSpec) { // activated pet
+        if (petMixer) {
+          // animated pet
+          if (petSpec) {
+            // activated pet
             const speed = 0.0014;
 
-            const distance = _getAppDistance();
+            const { distance, position } = _getAppDistance();
+
             const moveDelta = localVector;
             moveDelta.setScalar(0);
-            if (_isFar(distance)) { // handle rounding errors
-              const position = player.position.clone();
-              position.y = 0;
-              const direction = position.clone()
-                .sub(app.position)
-                .normalize();
+
+            if (_isFar(distance)) {
+              // handle rounding errors
+              const diff = position.sub(app.position);
+
+              if (Math.abs(diff.y) >= HEIGHT_FOLLOW_LIMIT_Y) {
+                diff.y = 0;
+              }
+
+              const direction = diff.normalize();
               const maxMoveDistance = distance - minDistance;
               const moveDistance = Math.min(speed * timeDiff, maxMoveDistance);
-              moveDelta.copy(direction)
-                .multiplyScalar(moveDistance);
+              moveDelta.copy(direction).multiplyScalar(moveDistance);
               app.position.add(moveDelta);
-              app.quaternion.slerp(localQuaternion.setFromUnitVectors(localVector2.set(0, 0, 1), direction), 0.1);
+              app.quaternion.slerp(

```

Comment:
There is no need to slerp anything here. This should be resolved by the abstracted shared character controller, based on follow and look settings on it. Quaternions should not be computed manually here.